adminapi/base.py

```diff
@@ -0,0 +1,450 @@
+        context.  We could try harder to make them hashable with a fallback
+        method, but that would lead them to considered as different objects
+        after they got an object_id.
+        """
+
+        return self.object_id
+
+    def __repr__(self):
+        parent_repr = super(BaseServerObject, self).__repr__()
+        if not self.object_id:
+            return 'ServerObject({0})'.format(parent_repr)
+        return 'ServerObject({0}, {1})'.format(parent_repr, self.object_id)
+
+    def is_dirty(self):
+        return bool(self.old_values) or self._deleted or self.object_id is None
+
+    def is_deleted(self):
+        return self._deleted
+
+    def commit(self):
+        raise NotImplementedError()
+
+    def rollback(self):
+        self._deleted = False
+        for attr, old_value in self.old_values.items():
+            super(BaseServerObject, self).__setitem__(attr, old_value)
+        self.old_values.clear()
+
+    def delete(self):
+        self._deleted = True
+
+    def _serialize_changes(self):
+        changes = {}
+        for key, old_value in self.old_values.items():
+            new_value = self[key]
+            if isinstance(old_value, MultiAttr):
+                action = 'multi'
+            else:
+                action = 'update'
+
+            change = {'action': action}
+            if action == 'update':
+                change['old'] = old_value
+                change['new'] = new_value
+            elif action == 'multi':
+                change['remove'] = old_value.difference(new_value)
+                change['add'] = new_value.difference(old_value)
+
+            changes[key] = change
+
+        return changes
+
+    def _confirm_changes(self):
+        self.old_values.clear()
+        if self._deleted:
+            self.object_id = None
+            self._deleted = False
+
+    def _build_commit_object(self):
+        changes = {}
+        if self.is_dirty():
+            changes[self.object_id] = self._serialize_changes()
+        return {
+            'deleted': [self.object_id] if self.is_deleted() else [],
+            'changes': changes,
+        }
+
+    def _save_old_value(self, key):
+        if key not in self.old_values:
+            old_value = self[key]
+            if isinstance(old_value, MultiAttr):
+                self.old_values[key] = old_value.copy()
+            else:
+                self.old_values[key] = old_value
+
+    def __setitem__(self, key, value):
+        if self._deleted:
+            raise DatasetError('Cannot set attributes to deleted server')
+        self.validate(key, value)
+        if isinstance(self[key], MultiAttr):
+            # Multi attributes are guaranteed to exist as MultiAttr, so
+            # we can always get the previous datatype from the existing
+            # value.
+            value = MultiAttr(value, self, key, self[key].datatype)
+        if key not in self.old_values or self.old_values[key] != value:
```

Comment:
I don't think we need that if clause here. Half of it is already in the `_save_old_value` and the other half does not really make sense for me. Just try to save it.

Reply:
We need to save the first version only if got set multiple times.

Reply:
On a second look, yes, it indeed doesn't make sense.